scene/resources/fog_material.cpp

```diff
@@ -0,0 +1,205 @@
+	ClassDB::bind_method(D_METHOD("set_height_offset", "height_offset"), &FogMaterial::set_height_offset);
+	ClassDB::bind_method(D_METHOD("get_height_offset"), &FogMaterial::get_height_offset);
+	ClassDB::bind_method(D_METHOD("set_height_falloff", "height_falloff"), &FogMaterial::set_height_falloff);
+	ClassDB::bind_method(D_METHOD("get_height_falloff"), &FogMaterial::get_height_falloff);
+	ClassDB::bind_method(D_METHOD("set_edge_fade", "edge_fade"), &FogMaterial::set_edge_fade);
+	ClassDB::bind_method(D_METHOD("get_edge_fade"), &FogMaterial::get_edge_fade);
+	ClassDB::bind_method(D_METHOD("set_density_texture", "density_texture"), &FogMaterial::set_density_texture);
+	ClassDB::bind_method(D_METHOD("get_density_texture"), &FogMaterial::get_density_texture);
+
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "density", PROPERTY_HINT_RANGE, "0.0,16.0,0.01,or_greator,or_lesser"), "set_density", "get_density");
+	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "albedo", PROPERTY_HINT_COLOR_NO_ALPHA), "set_albedo", "get_albedo");
+	ADD_PROPERTY(PropertyInfo(Variant::COLOR, "emission", PROPERTY_HINT_COLOR_NO_ALPHA), "set_emission", "get_emission");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "height_offset", PROPERTY_HINT_RANGE, "-128.0,128.0,0.01,or_greater,or_lesser"), "set_height_offset", "get_height_offset");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "height_falloff", PROPERTY_HINT_EXP_EASING, "attenuation"), "set_height_falloff", "get_height_falloff");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "edge_fade", PROPERTY_HINT_EXP_EASING), "set_edge_fade", "get_edge_fade");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "density_texture", PROPERTY_HINT_RESOURCE_TYPE, "Texture3D"), "set_density_texture", "get_density_texture");
+}
+
+void FogMaterial::cleanup_shader() {
+	if (shader.is_valid()) {
+		RS::get_singleton()->free(shader);
+	}
+}
+
+void FogMaterial::_update_shader() {
+	shader_mutex.lock();
+	if (shader.is_null()) {
+		shader = RS::get_singleton()->shader_create();
+
+		// Add a comment to describe the shader origin (useful when converting to ShaderMaterial).
+		RS::get_singleton()->shader_set_code(shader, R"(
+// NOTE: Shader automatically converted from )" VERSION_NAME " " VERSION_FULL_CONFIG R"('s FogMaterial.
+
+shader_type fog;
+
+uniform float density = 1.0;
+uniform vec4 albedo = vec4(1.0);
+uniform vec4 emission = vec4(0.0);
+uniform float height_offset = 0.0;
+uniform float height_falloff = 0.0;
+uniform float edge_fade = 0.1;
+uniform sampler3D density_texture: hint_white;
+
+
+void fog() {
+    DENSITY = density * exp2(-height_falloff * (WORLD_POSITION.y - height_offset));
+    DENSITY *= texture(density_texture, UVW).r;
+    DENSITY *= pow(clamp(-SDF / min(min(EXTENTS.x, EXTENTS.y), EXTENTS.z), 0.0, 1.0), edge_fade);
+    ALBEDO = albedo.rgb;
+    EMISSION = emission.rgb;
+}
+)");
+	}
+	shader_mutex.unlock();
+}
+
+FogMaterial::FogMaterial() {
+	set_density(1.0);
+	set_albedo(Color(1, 1, 1, 1));
+	set_emission(Color(0, 0, 0, 0));
```

Comment:
since emission only uses rgb (unless changed in the last hour), I think it makes sense to make alpha=1 if just for the sake of the color being visible in the inspector

Reply:
```suggestion
	set_emission(Color(0, 0, 0, 1));
```